### PR TITLE
Port lunar lander to pymunk

### PR DESCRIPTION
**Pull request by Lonny154**
# Description

This draft PR adds a standalone proof of concept for representing the core `LunarLander` physics with Pymunk instead of Box2D.

The prototype currently demonstrates:

* reproducible seeded terrain generation with a flat helipad
* a Pymunk lander hull
* two constrained landing legs
* independent leg-to-terrain contact tracking
* hull-to-terrain crash detection
* main-engine thrust
* orientation-engine thrust in both directions
* a small discrete action interface
* invariant-based tests for the prototype behavior

The existing Box2D-backed `LunarLander` environment, environment registrations, dependencies, documentation, and public API are unchanged.

This is intended to demonstrate feasibility and gather feedback on the migration approach.

Related to #1597

## Type of change

* [ ] Documentation only change (no code changed)
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Testing

The added tests verify that:

* terrain generation is reproducible for a fixed seed
* generated state values remain finite
* main-engine thrust reduces downward velocity relative to no thrust
* the two orientation actions produce opposite angular acceleration
* the legs register contact after a normal landing
* hull contact sets the crash flag

Commands run:

```bash
pytest tests/envs/test_pymunk_lunar_lander_terrain.py
```

Result:

```text
6 passed
```

The pre-commit checks were also run successfully on the files changed by this PR:

```bash
pre-commit run --files \
  scripts/pymunk_lunar_lander_terrain.py \
  tests/envs/test_pymunk_lunar_lander_terrain.py
```

# Checklist

* [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
* [x] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by jkterry1** [comment]
@Lonny154 are you able to show before/after learning performance graphs?

**Comment by Lonny154** [comment]
Not yet, I am working on that now.

**Comment by Lonny154** [comment]
I got an experimental Pymunk environment to run for LunarLander. Here is a training graph comparing box2d and pymunk. 

image


One Pymunk seed reached approximately 198 mean return and a 90% success rate at 300,000 steps, while the other two seeds remained substantially weaker. The Box2D runs also showed high seed sensitivity and did not consistently solve the environment under this PPO configuration.

**Comment by jkterry1** [comment]
-Are there any visual indications of physics differences that might cause what we're seeing when doing the rollouts?
-I'm not sure that 3 training runs is a large enough sample size for something of this nater

-For what it's worth for future context I'm not willing to merge these PRs myself, I'd only feel comfortable with @pseudo-rnd-thoughts doing so

**Comment by Lonny154** [comment]
Yeah, I do not think this PR is ready to be merged yet. It is still a rough draft, and the additional seeded runs and visual comparisons have made it clear that there are physics differences that need further investigation. 

After working through these initial phases, I have realized that the process is more involved than I initially anticipated. I would greatly enjoy continuing to work through the transition, but I think doing it properly would realistically be a multi-month effort.

**Comment by jkterry1** [comment]
Hey,

I completely understand. Is relatively quickly creating initial draft ports of the other box2d environments and leaving them up for future travelers to start with and debug into ready to merge implementations something you might be interested in?

-------- Original Message --------

**Comment by Lonny154** [comment]
Yea that would be something I am interested in!

**Comment by jkterry1** [comment]
That would be extremely helpful, thank you :)

-------- Original Message --------

**Comment by pseudo-rnd-thoughts** [comment]
Thanks for the PR @Lonny154 

@jkterry1 We need to set some standards for when we feel comfortable to merge this. We can't exact equivalent behaviour but I think we need to have
1. Same goal and reward functions
2. Similar training curves for the same algorithms + hyperparameters (I think SB3 or CleanRL PPO, SAC and DQN would be good testing)
3. Higher implementation standards as backward compatibility doesn't exist yet

Anything else? 

For environment versions, I think we could imagine it like Mujoco-py to Mujoco where we can keep both implementations side-by-side for a while and this version being the latest version

**Comment by jkterry1** [comment]
@pseudo-rnd-thoughts That all sounds good to me, with the note that I was already suggesting keeping both versions side by side for awhile in the original issue asking for these three (with warnings on each that one is being yanked and the other one is being tested and please report bugs/differences for awhile)

**Comment by Lonny154** [comment]
Hello, I’m still working through this and have made some progress. I added fixed-action physics comparisons to separate learned-policy behavior from environment dynamics. A few of the things I was able to diagnose and fix with those were reset behavior and side engine dynamics.

**Comment by jkterry1** [comment]
@Lonny154 could you please rebase off main?

**Comment by Lonny154** [comment]
Rebased this branch onto the latest main and force-pushed the updated history.

I have also completed a more detailed Box2D/Pymunk parity investigation. The main corrections include:

- Matching reset-state randomness and engine dispersion
- Matching hull and leg mass properties, hull center of gravity, engine impulse magnitudes, and engine application points
- Correcting leg geometry, joint configuration, terrain collision radius, and effective contact friction
- Adding Box2D-style sleeping tolerances and a deterministic stability fallback for successful landing termination
- Auditing reward decomposition and confirming that the reconstructed reward components match observed episode returns

With the same DQN hyperparameters across five seeds, both implementations reached approximately 250 mean return around 250k–300k steps. After 300k steps, Box2D averaged 239.9 return with an 83.3% stable-landing rate, while Pymunk averaged 256.5 with a 96.2% stable-landing rate.

PPO showed more seed sensitivity, but both implementations improved on a similar timescale using identical settings. At 300k steps, Box2D averaged 128.8 return and Pymunk 63.9, with both reaching a 25% return-threshold success rate. PPO update diagnostics, including approximate KL divergence, clipping fraction, policy-gradient loss, and explained variance, did not indicate an optimizer instability specific to Pymunk. The remaining difference was primarily behavioral: Pymunk policies selected a hovering/time-limit strategy more frequently.

Pre-commit now passes on the rebased branch, and the focused Pymunk environment and comparison-script tests pass. I would appreciate feedback on whether the current PPO and DQN evidence is sufficient for this stage, or whether a continuous-action SAC comparison should be the next benchmark.

**Comment by Lonny154** [comment]
I've attached a technical report that goes into more detail on what I updated in the most recent commits. 

[LunarLander_Technicalreport.pdf](https://github.com/user-attachments/files/30129511/LunarLander_Technicalreport.pdf)

**Comment by jkterry1** [comment]
@Lonny154 what are your next steps as port of this porting work?

**Comment by Lonny154** [comment]
I went through the recent Multiwalker Pymunk parity findings to see whether any of the same issues applied to LunarLander.

I found one real semantic mismatch: the Pymunk prototype had an extra hard-leg-impact termination heuristic that Box2D does not have. I removed that and added a direct Box2D/Pymunk regression test.

I also validated the leg motor torque mapping with isolated loaded-joint experiments across multiple loads. `maxMotorTorque=40` in Box2D matched `max_force=40` in Pymunk closely, while scaling by `dt` was clearly incorrect.

I’ve also updated the branch with the latest `main`, and the LunarLander tests are still passing. My next step would be to rerun the PPO/DQN learning comparisons now that the remaining parity issues I found are addressed. After that, I’d be interested in feedback on whether there are any other LunarLander-specific gaps you’d like me to investigate or whether I should move on to another environment.

**Comment by jkterry1** [comment]
Awesome, can you please rerun those comparison tests?

-------- Original Message --------

**Comment by Lonny154** [comment]
I reran the fixed-action, PPO, and DQN comparisons. Fixed-action parity is still close, PPO is broadly comparable, but DQN still shows a meaningful behavioral difference. I’m going to treat the DQN gap and code-quality cleanup as the main remaining work, then get the implementation into the best merge-ready state I can and push the update

**Comment by Lonny154** [comment]
## LunarLander Pymunk acceptance results

PPO, DQN, and SAC satisfy the benchmark criterion used for this migration. Across the paired training seeds, I found no statistically supported systematic performance deficit for Pymunk.

Following the discussion in #1597 and this PR, I operationalized the requested “similar training curves” standard as:

* Identical algorithms and hyperparameters across engines
* 10 paired training seeds
* 500,000 training steps per engine and seed
* Evaluation every 10,000 steps
* 20 shared, fixed evaluation seeds per checkpoint
* Seed-level paired bootstrap confidence intervals using 10,000 samples
* No requirement for exact trajectory or point-estimate equality

Candidate commit: `9f8cb62e6b4f5b1f48c5f157672c9c08b7ac2533`

Source fingerprint: `35f29ef83d022f6782217f83315d6942cfb2be81df545019aaef8617cfbb53f5`

### PPO — Pass

* Final return: Box2D `130.4`, Pymunk `121.2`
* Landing rate: Box2D `68.2%`, Pymunk `62.2%`
* Paired final-return difference: `+9.2`, 95% CI `[-35.8, 56.1]`
* Paired normalized-AUC difference: `-58.1`, 95% CI `[-138.2, 13.1]`
* Engine uncertainty overlaps throughout training.

### DQN — Pass with qualifications

* Final return: Box2D `152.0`, Pymunk `108.1`
* Landing rate: Box2D `61.3%`, Pymunk `57.3%`
* Paired final-return difference: `+43.9`, 95% CI `[-73.1, 219.4]`
* Paired normalized-AUC difference: `+18.6`, 95% CI `[-42.5, 69.2]`
* Both engines show substantial seed and late-training instability.
* Pymunk trails around 400k–460k steps, but the gap narrows by 500k.

### SAC — Pass with qualifications

* Final return: Box2D `194.7`, Pymunk `195.9`
* Landing rate: Box2D `70.2%`, Pymunk `69.5%`
* Paired final-return difference: `-1.2`, 95% CI `[-87.2, 79.4]`
* Paired normalized-AUC difference: `+12.5`, 95% CI `[-65.4, 85.5]`
* Both engines show bimodal seed convergence.
* Final aggregate return and landing performance are nearly identical.

All paired differences are calculated as `Box2D - Pymunk`. Every paired 95% confidence interval includes zero.

PPO and DQN compare `LunarLander-v3` with `LunarLander-v4`. SAC compares `LunarLanderContinuous-v3` with `LunarLanderContinuous-v4`. Both backends use the same `TimeLimit(1000)` contract and matching action spaces and environment settings within each algorithm.

All three schema-5 manifests:

* Record the frozen candidate and clean-source fingerprint
* Completed all 20 expected engine/seed pairs
* Contain checkpoints from 0 through 500,000 steps
* Contain complete and unique aggregate and episode keys
* Use the same training and evaluation seeds across engines
* Reproduce their generated summaries from the raw rows

The audit covered 3,060 aggregate checkpoint rows and 61,200 evaluation episodes. Every outcome partition sums to one within floating-point precision, and every reward-decomposition residual is below the documented `1e-3` tolerance.

These results support comparable learning performance for the Pymunk environments. They do not establish exact physical, trajectory, or statistical equivalence between Pymunk and Box2D.

### Learning curves
curve
curve
curve

**Comment by pseudo-rnd-thoughts** [comment]
I'm skim read your results @Lonny154 and they look great. I'll cut the 1.4.0 version tomorrow then have a serious look at the solution and hope to merge it for 1.5.0

**Comment by Lonny154** [comment]
Awesome sounds good! @pseudo-rnd-thoughts
